contributors/design-proposals/mount-options.md

```diff
@@ -0,0 +1,145 @@
+# Mount options for mountable volume types
+
+## Goal
+
+Enable Kubernetes users to specify mount options with mountable volumes
+such as  - `nfs`, `glusterfs` or `aws-ebs` etc.
+
+## Motivation
+
+We currently support network filesystems: NFS, Glusterfs, Ceph FS, SMB (Azure file), Quobytes, and local filesystems such as ext[3|4] and XFS.
+
+Mount time options that are operationally important and have no security implications should be suppported. Examples are NFS's TCP mode, versions, lock mode, caching mode; Glusterfs's caching mode; SMB's version, locking, id mapping; and more.
+
+## Design
+
+Mount options can be specified as `mountOption`  as part of PV. As an example:
+
+``` yaml
+ apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    name: pv0003
+  spec:
+    capacity:
+      storage: 5Gi
+    accessModes:
+      - ReadWriteOnce
+    persistentVolumeReclaimPolicy: Recycle
+    nfs:
+      path: /tmp
+      server: 172.17.0.2
+      mountOption: "hard,nolock,nfsvers=3"
+```
+
+To be consistent with VolumeSource in pod specs, it can also be specified as part of Volume spec with Pod:
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pd
+spec:
+  containers:
+  - image: gcr.io/google_containers/test-webserver
+    name: test-container
+    volumeMounts:
+    - mountPath: /test-pd
+      name: test-volume
+  volumes:
+  - name: test-volume
+    nfs:
+      server: "192.168.1.100"
+      path: "/var/exports"
+      mountOption: "hard,nolock,nfsvers=3"
+```
+
+Or it can be specified as part of StorageClass definition:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: StorageClass
+metadata:
+  name: slow
+provisioner: kubernetes.io/glusterfs
+parameters:
+  resturl: "http://127.0.0.1:8081"
+  restauthenabled: "true"
+  restuser: "admin"
+  secretNamespace: "default"
+  secretName: "heketi-secret"
+  mountOption: "volume-name=foobar"
+```
+
+## Preventing users from specifying invalid/unsupported mount options
+
+While mount options enable more flexibility in how volumes are mounted, it can result
+in user specifying options that are not supported or are known to be problematic.
+
+There could be several possibly alternatives for this. I am listing few I considered.
+
+
+1. Add support for blacklisting mount options as `kubelet` startup configuration:
+
+To prevent users from using options which are not safe - Kubernetes admin can configure volume options and specify a blacklist of mount point parameters. For example:
+
+```
+~> kubelet --volume-options <volume-config-file>
```

Comment:
+1 let's not make the white/black list cluster admin configurable. Just hard code it in the plugins for now. The addition of the `mountOptions` field to the volume specs for some plugins should be the only API change.

My thought process for this is that we should minimize changes to the API while making sure we don't paint ourselves in to a corner. And by implementing a minimally viable solution (hardcoded white/black list), we reduce complexity while leaving open the option of adding in cluster admin configurable list in the future if the need arises.